File: anagrams.java
import java.io.File;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;

import java.util.Arrays;

class anagram {
  public static void main(String args[]) {
    try{
      File file = new File("/usr/share/dict/words");
      BufferedReader dicReader = new BufferedReader(new FileReader(file));
      BufferedReader strReader = new BufferedReader(new InputStreamReader(System.in), 1);

      String str = strReader.readLine();
      String strSort = new String(sort(str));
      strReader.close();

      String dic;
      String ans = "";
      while((dic = dicReader.readLine()) != null) {
        if(isAnagram(strSort, dic) && (dic.length() > ans.length())) ans = dic;
      }
      System.out.println(ans);
      dicReader.close();
    } catch(FileNotFoundException e){
      System.out.println(e);
    } catch (IOException e){
      System.out.println(e);
    }
  }

  public static char[] sort(String str) {
    char[] chars = str.toLowerCase().toCharArray();
    Arrays.sort(chars);
    return chars;
  }

  public static boolean isAnagram(String str, String dic) {
    char[] dicSorts = sort(dic);
    int pos = -1;
    for(int i = 0; i < dicSorts.length; i++){
      pos = str.indexOf(dicSorts[i], pos+1);
      if(pos == -1) return false;
    }
    return true;
  }
}

File: anagrams.rs
use std::error::Error;
use std::fs::File;
use std::io;
use std::io::BufReader;
use std::io::prelude::*;
use std::path::Path;

fn sort(s: &str) -> String {
    let mut chars: Vec<_> = s.chars().collect();
    chars.sort();
    chars.into_iter().collect()
}

fn check_anagram(target: &str, word: &str) -> bool {
    let sorted = sort(word);
    let mut pos = 0;
    for c in sorted.chars() {
        pos = match target[pos..].find(c) {
            None => return false,
            Some(p) => p + pos + 1,
        };
    }
    true
}

fn main() {
    let mut ans = String::new();
    let mut target = String::new();
    io::stdin().read_line(&mut target)
        .expect("failed to read line");

    let target: String = match target.trim().parse() {
        Ok(s) => s,
        Err(_) => panic!("couldn't read stdin."),
    };
    
    let sorted_target = sort(&target);

    let path = Path::new("/usr/share/dict/words");
    let display = path.display();

    let file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", display,
                           Error::description(&why)),
        Ok(file) => file,
    };

    let reader = BufReader::new(file);
    let lines = reader.lines().map(|l| l.unwrap());

    for word in lines {
        let is_anagram = check_anagram(&sorted_target, &word);
        if is_anagram && (word.len() > ans.len()) {
            ans = word;
        }
    }
    println!("{} is the longest anagram of {}.", ans, target);
}
